File: app/src/main/java/no/joharei/flixr/CardPresenter.java
package no.joharei.flixr;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.v17.leanback.widget.ImageCardView;
import android.support.v17.leanback.widget.Presenter;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.ViewGroup;

import com.squareup.picasso.Picasso;

import no.joharei.flixr.api.models.Photo;
import no.joharei.flixr.api.models.Photoset;
import no.joharei.flixr.mainpage.models.Contact;

public class CardPresenter extends Presenter {
    private static final String TAG = "CardPresenter";

    private static final int CARD_HEIGHT = 176;
    private static int sSelectedBackgroundColor;
    private static int sDefaultBackgroundColor;
    private Drawable mDefaultCardImage;

    private static void updateCardBackgroundColor(ImageCardView view, boolean selected) {
        int color = selected ? sSelectedBackgroundColor : sDefaultBackgroundColor;
        // Both background colors should be set because the view's background is temporarily visible
        // during animations.
        view.setBackgroundColor(color);
        view.findViewById(R.id.info_field).setBackgroundColor(color);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent) {
        final Context context = parent.getContext();
        sDefaultBackgroundColor = ContextCompat.getColor(context, R.color.default_background);
        sSelectedBackgroundColor = ContextCompat.getColor(context, R.color.selected_background);
        mDefaultCardImage = ContextCompat.getDrawable(context, R.drawable.movie);

        ImageCardView cardView = new ImageCardView(context) {
            @Override
            public void setSelected(boolean selected) {
                updateCardBackgroundColor(this, selected);
                super.setSelected(selected);
            }
        };

        cardView.setFocusable(true);
        cardView.setFocusableInTouchMode(true);
        updateCardBackgroundColor(cardView, false);
        return new ViewHolder(cardView);
    }

    @Override
    public void onBindViewHolder(Presenter.ViewHolder viewHolder, Object item) {
        ImageCardView cardView = (ImageCardView) viewHolder.view;

        cardView.setMainImageDimensions(ViewGroup.LayoutParams.WRAP_CONTENT, CARD_HEIGHT);
        Context context = viewHolder.view.getContext();
        if (item instanceof Photoset) {
            Photoset photoset = (Photoset) item;
            cardView.setTitleText(photoset.getTitle().getContent());
            cardView.setContentText(photoset.getDescription().getContent());
            Picasso.with(context)
                    .load(photoset.getCardImageUrl())
                    .error(mDefaultCardImage)
                    .into(cardView.getMainImageView());
        } else if (item instanceof Photo) {
            Photo photo = (Photo) item;
            cardView.setTitleText(photo.getTitle());
            Picasso.with(context)
                    .load(photo.getCardImageUrl())
                    .error(mDefaultCardImage)
                    .into(cardView.getMainImageView());
        } else if (item instanceof Contact) {
            Contact contact = (Contact) item;
            cardView.setTitleText(!(contact.getRealname() == null || contact.getRealname().isEmpty()) ?
                    contact.getRealname() :
                    contact.getUsername());
            Log.d(TAG, "Contact photo: " + contact.getCardImageUrl());
            Picasso.with(context)
                    .load(contact.getCardImageUrl())
                    .error(mDefaultCardImage)
                    .into(cardView.getMainImageView());
        }
    }

    @Override
    public void onUnbindViewHolder(Presenter.ViewHolder viewHolder) {
        ImageCardView cardView = (ImageCardView) viewHolder.view;
        // Remove references to images so that the garbage collector can free up memory
        cardView.setBadgeImage(null);
        cardView.setMainImage(null);
    }
}

File: app/src/main/java/no/joharei/flixr/CardPresenter.kt
package no.joharei.flixr

import android.content.Context
import android.text.TextUtils
import android.view.ContextThemeWrapper
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.leanback.widget.ImageCardView
import androidx.leanback.widget.Presenter
import no.joharei.flixr.common.extensions.dimen
import no.joharei.flixr.glide.GlideApp
import no.joharei.flixr.mainpage.models.Contact
import no.joharei.flixr.network.models.Photoset

class CardPresenter(val context: Context) : Presenter() {
    private var sSelectedBackgroundColor: Int = 0
    private var sDefaultBackgroundColor: Int = 0
    private lateinit var cardView: ImageCardView
    private val imageHeight: Int by lazy { context.dimen(R.dimen.photo_card_height) }

    override fun onCreateViewHolder(parent: ViewGroup): Presenter.ViewHolder {
        val context = parent.context
        sDefaultBackgroundColor = ContextCompat.getColor(context, R.color.default_background)
        sSelectedBackgroundColor = ContextCompat.getColor(context, R.color.selected_background)

        val cardView =
            object : ImageCardView(ContextThemeWrapper(context, R.style.CustomImageCardTheme)) {
                val textView = findViewById<TextView>(R.id.title_text)

                init {
                    isFocusable = true
                    isFocusableInTouchMode = true
                    setMainImageDimensions(WRAP_CONTENT, imageHeight)
                }

                override fun setSelected(selected: Boolean) {
                    updateCardBackgroundColor(this, selected)
                    textView.ellipsize =
                            if (selected) TextUtils.TruncateAt.MARQUEE else TextUtils.TruncateAt.END
                    textView.setHorizontallyScrolling(selected)
                    super.setSelected(selected)
                }
            }

        updateCardBackgroundColor(cardView, false)
        return Presenter.ViewHolder(cardView)
    }

    override fun onBindViewHolder(viewHolder: Presenter.ViewHolder, item: Any) {
        cardView = viewHolder.view as ImageCardView
        cardView.titleText = when (item) {
            is Photoset -> item.title
            is Contact -> item.displayName
            else -> null
        }
        when (item) {
            is Photoset -> scaleWidth(item.thumbnailHeight, item.thumbnailWidth, imageHeight)
            is Contact -> scaleWidth(item.thumbnailDim, item.thumbnailDim, imageHeight)
            else -> null
        }?.let {
            cardView.setMainImageDimensions(it, imageHeight)
        }
        GlideApp.with(context)
            .load(
                when (item) {
                    is Photoset -> item.photoUrl(
                        Int.MAX_VALUE,
                        context.dimen(R.dimen.photo_card_height)
                    )
                    is Contact -> item.cardImageUrl
                    else -> null
                }
            )
            .placeholder(R.color.black_opaque)
            .error(R.drawable.ic_error)
            .into(cardView.mainImageView)
    }

    private fun scaleWidth(originalHeight: Int, originalWidth: Int, newHeight: Int) =
        newHeight * originalWidth / originalHeight

    override fun onUnbindViewHolder(viewHolder: Presenter.ViewHolder) {
        val cardView = viewHolder.view as ImageCardView
        // Remove references to images so that the garbage collector can free up memory
        cardView.badgeImage = null
        cardView.mainImage = null
    }

    private fun updateCardBackgroundColor(view: ImageCardView, selected: Boolean) {
        val color = if (selected) sSelectedBackgroundColor else sDefaultBackgroundColor
        // Both background colors should be set because the view's background is temporarily visible
        // during animations.
        view.setBackgroundColor(color)
        view.findViewById<View>(R.id.info_field).setBackgroundColor(color)
    }
}
